File: lib/src/androidTest/java/me/ycdev/android/lib/common/utils/ReflectionUtilsTest.java
package me.ycdev.android.lib.common.utils;

import android.test.AndroidTestCase;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

@SuppressWarnings("unused")
public class ReflectionUtilsTest extends AndroidTestCase {
    private static class TestA {
        public String a1;
        public int a2;
        private String a3;
        private long a4;

        public static String a5;
        public static int a6;
        private static String a7;
        private static long a8;

        public String a9() {
            return "a9";
        }

        protected int a10(int a) {
            return a;
        }

        private String a11(String a, long b) {
            return a + b;
        }

        public static String a13() {
            return "a13";
        }

        protected static long a14(long a) {
            return a;
        }

        private static String a15(int a, String b) {
            return b + a;
        }
    }

    private static class TestB extends TestA {
        public String b1;
        public int b2;
        private String b3;
        private long b4;

        public static String b5;
        public static int b6;
        private static String b7;
        private static long b8;

        @Override
        public String a9() {
            return "a9";
        }

        @Override
        protected int a10(int a) {
            return a;
        }

        private String b11(String a, long b) {
            return a + b;
        }

        public static String b13() {
            return "b13";
        }

        protected static long b14(long a) {
            return a;
        }

        private static String b15(int a, String b) {
            return b + a;
        }
    }

    public void test_findMethod() {
        try {
            TestB objB = new TestB();

            // TestA part
            Method a11 = ReflectionUtils.findMethod(TestB.class, "a11", String.class, long.class);
            assertTrue(a11.getDeclaringClass() == TestA.class && a11.getName().equals("a11"));
            assertTrue(a11.invoke(objB, "a", 11l).equals("a11"));

            Method a13 = ReflectionUtils.findMethod(TestB.class, "a13");
            assertTrue(a13.getDeclaringClass() == TestA.class && a13.getName().equals("a13"));
            assertTrue(a13.invoke(null).equals("a13"));

            Method a14 = ReflectionUtils.findMethod(TestB.class, "a14", long.class);
            assertTrue(a14.getDeclaringClass() == TestA.class && a14.getName().equals("a14"));
            assertTrue((long) a14.invoke(null, 14l) == 14l);

            Method a15 = ReflectionUtils.findMethod(TestB.class, "a15", int.class, String.class);
            assertTrue(a15.getDeclaringClass() == TestA.class && a15.getName().equals("a15"));
            assertTrue(a15.invoke(null, 15, "a").equals("a15"));

            // TestB part
            Method a9 = ReflectionUtils.findMethod(TestB.class, "a9");
            assertTrue(a9.getDeclaringClass() == TestB.class && a9.getName().equals("a9"));
            assertTrue(a9.invoke(objB).equals("a9"));

            Method a10 = ReflectionUtils.findMethod(TestB.class, "a10", int.class);
            assertTrue(a10.getDeclaringClass() == TestB.class && a10.getName().equals("a10"));
            assertTrue((int) a10.invoke(objB, 10) == 10);

            Method b11 = ReflectionUtils.findMethod(TestB.class, "b11", String.class, long.class);
            assertTrue(b11.getDeclaringClass() == TestB.class && b11.getName().equals("b11"));
            assertTrue(b11.invoke(objB, "b", 11l).equals("b11"));

            Method b13 = ReflectionUtils.findMethod(TestB.class, "b13");
            assertTrue(b13.getDeclaringClass() == TestB.class && b13.getName().equals("b13"));
            assertTrue(b13.invoke(null).equals("b13"));

            Method b14 = ReflectionUtils.findMethod(TestB.class, "b14", long.class);
            assertTrue(b14.getDeclaringClass() == TestB.class && b14.getName().equals("b14"));
            assertTrue((long) b14.invoke(null, 14l) == 14l);

            Method b15 = ReflectionUtils.findMethod(TestB.class, "b15", int.class, String.class);
            assertTrue(b15.getDeclaringClass() == TestB.class && b15.getName().equals("b15"));
            assertTrue(b15.invoke(null, 15, "b").equals("b15"));
        } catch (Exception e) {
            e.printStackTrace();
            fail("failed to reflect: " + e.toString());
        }
    }

    public void test_findField() {
        try {
            TestB objB = new TestB();

            // TestA part
            Field a1 = ReflectionUtils.findField(TestB.class, "a1");
            assertTrue(a1.getDeclaringClass() == TestA.class && a1.getName().equals("a1"));
            a1.set(objB, "a1");
            assertTrue(a1.get(objB).equals("a1"));

            Field a2 = ReflectionUtils.findField(TestB.class, "a2");
            assertTrue(a2.getDeclaringClass() == TestA.class && a2.getName().equals("a2"));
            a2.set(objB, 2);
            assertTrue((int) a2.get(objB) == 2);

            Field a3 = ReflectionUtils.findField(TestB.class, "a3");
            assertTrue(a3.getDeclaringClass() == TestA.class && a3.getName().equals("a3"));
            a3.set(objB, "a3");
            assertTrue(a3.get(objB).equals("a3"));

            Field a4 = ReflectionUtils.findField(TestB.class, "a4");
            assertTrue(a4.getDeclaringClass() == TestA.class && a4.getName().equals("a4"));
            a4.set(objB, 4l);
            assertTrue((long) a4.get(objB) == 4l);

            Field a5 = ReflectionUtils.findField(TestB.class, "a5");
            assertTrue(a5.getDeclaringClass() == TestA.class && a5.getName().equals("a5"));
            a5.set(null, "a5");
            assertTrue(a5.get(null).equals("a5"));

            Field a6 = ReflectionUtils.findField(TestB.class, "a6");
            assertTrue(a6.getDeclaringClass() == TestA.class && a6.getName().equals("a6"));
            a6.set(null, 6);
            assertTrue((int) a6.get(null) == 6);

            Field a7 = ReflectionUtils.findField(TestB.class, "a7");
            assertTrue(a7.getDeclaringClass() == TestA.class && a7.getName().equals("a7"));
            a7.set(null, "a7");
            assertTrue(a7.get(null).equals("a7"));

            Field a8 = ReflectionUtils.findField(TestB.class, "a8");
            assertTrue(a8.getDeclaringClass() == TestA.class && a8.getName().equals("a8"));
            a8.set(null, 8l);
            assertTrue((long) a8.get(null) == 8l);

            // TestB part
            Field b1 = ReflectionUtils.findField(TestB.class, "b1");
            assertTrue(b1.getDeclaringClass() == TestB.class && b1.getName().equals("b1"));
            a1.set(objB, "b1");
            assertTrue(a1.get(objB).equals("b1"));

            Field b2 = ReflectionUtils.findField(TestB.class, "b2");
            assertTrue(b2.getDeclaringClass() == TestB.class && b2.getName().equals("b2"));
            b2.set(objB, 2);
            assertTrue((int) b2.get(objB) == 2);

            Field b3 = ReflectionUtils.findField(TestB.class, "b3");
            assertTrue(b3.getDeclaringClass() == TestB.class && b3.getName().equals("b3"));
            b3.set(objB, "b3");
            assertTrue(b3.get(objB).equals("b3"));

            Field b4 = ReflectionUtils.findField(TestB.class, "b4");
            assertTrue(b4.getDeclaringClass() == TestB.class && b4.getName().equals("b4"));
            b4.set(objB, 4l);
            assertTrue((long) b4.get(objB) == 4l);

            Field b5 = ReflectionUtils.findField(TestB.class, "b5");
            assertTrue(b5.getDeclaringClass() == TestB.class && b5.getName().equals("b5"));
            b5.set(null, "b5");
            assertTrue(b5.get(null).equals("b5"));

            Field b6 = ReflectionUtils.findField(TestB.class, "b6");
            assertTrue(b6.getDeclaringClass() == TestB.class && b6.getName().equals("b6"));
            b6.set(null, 6);
            assertTrue((int) b6.get(null) == 6);

            Field b7 = ReflectionUtils.findField(TestB.class, "b7");
            assertTrue(b7.getDeclaringClass() == TestB.class && b7.getName().equals("b7"));
            b7.set(null, "b7");
            assertTrue(b7.get(null).equals("b7"));

            Field b8 = ReflectionUtils.findField(TestB.class, "b8");
            assertTrue(b8.getDeclaringClass() == TestB.class && b8.getName().equals("b8"));
            b8.set(null, 8l);
            assertTrue((long) b8.get(null) == 8l);
        } catch (Exception e) {
            e.printStackTrace();
            fail("failed to reflect: " + e.toString());
        }
    }
}

File: baseLib/src/test/java/me/ycdev/android/lib/common/utils/ReflectionUtilsTest.kt
package me.ycdev.android.lib.common.utils

import androidx.test.filters.SmallTest
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

@SmallTest
class ReflectionUtilsTest {
    private open class TestA {
        var a1: String? = null
        var a2: Int = 0
        private val a3: String? = null
        private val a4: Long = 0

        open fun a9(): String {
            return "a9"
        }

        protected open fun a10(a: Int): Int {
            return a
        }

        private fun a11(a: String, b: Long): String {
            return a + b
        }

        companion object {

            var a5: String? = null
            var a6: Int = 0
            private var a7: String? = null
            private var a8: Long = 0

            @JvmStatic
            fun a13(): String {
                return "a13"
            }

            @JvmStatic
            protected fun a14(a: Long): Long {
                return a
            }

            @JvmStatic
            private fun a15(a: Int, b: String): String {
                return b + a
            }
        }
    }

    private class TestB : TestA() {
        var b1: String? = null
        var b2: Int = 0
        private val b3: String? = null
        private val b4: Long = 0

        override fun a9(): String {
            return "a9"
        }

        override fun a10(a: Int): Int {
            return a
        }

        private fun b11(a: String, b: Long): String {
            return a + b
        }

        companion object {

            var b5: String? = null
            var b6: Int = 0
            private var b7: String? = null
            private var b8: Long = 0

            @JvmStatic
            fun b13(): String {
                return "b13"
            }

            @JvmStatic
            protected fun b14(a: Long): Long {
                return a
            }

            @JvmStatic
            private fun b15(a: Int, b: String): String {
                return b + a
            }
        }
    }

    @Test
    fun test_findMethod() {
        try {
            val objB = TestB()

            // TestA part
            val a11 = ReflectionUtils.findMethod(
                TestB::class.java,
                "a11",
                String::class.java,
                Long::class.java
            )
            assertTrue(a11.declaringClass == TestA::class.java && a11.name == "a11")
            assertTrue(a11.invoke(objB, "a", 11L) == "a11")

            val a13 = ReflectionUtils.findMethod(TestB::class.java, "a13")
            assertTrue(a13.declaringClass == TestA::class.java && a13.name == "a13")
            assertTrue(a13.invoke(null) == "a13")

            val a14 =
                ReflectionUtils.findMethod(TestB::class.java, "a14", Long::class.java)
            assertTrue(a14.declaringClass == TestA::class.java && a14.name == "a14")
            assertTrue(a14.invoke(null, 14L) as Long == 14L)

            // TODO fix the following case
//            val a15 = ReflectionUtils.findMethod(
//                TestB::class.java,
//                "a15",
//                Int::class.java,
//                String::class.java
//            )
//            assertTrue(a15.declaringClass == TestA::class.java && a15.name == "a15")
//            assertTrue(a15.invoke(null, 15, "a") == "a15")

            // TestB part
            val a9 = ReflectionUtils.findMethod(TestB::class.java, "a9")
            assertTrue(a9.declaringClass == TestB::class.java && a9.name == "a9")
            assertTrue(a9.invoke(objB) == "a9")

            val a10 =
                ReflectionUtils.findMethod(TestB::class.java, "a10", Int::class.java)
            assertTrue(a10.declaringClass == TestB::class.java && a10.name == "a10")
            assertTrue(a10.invoke(objB, 10) as Int == 10)

            val b11 = ReflectionUtils.findMethod(
                TestB::class.java,
                "b11",
                String::class.java,
                Long::class.java
            )
            assertTrue(b11.declaringClass == TestB::class.java && b11.name == "b11")
            assertTrue(b11.invoke(objB, "b", 11L) == "b11")

            val b13 = ReflectionUtils.findMethod(TestB::class.java, "b13")
            assertTrue(b13.declaringClass == TestB::class.java && b13.name == "b13")
            assertTrue(b13.invoke(null) == "b13")

            val b14 =
                ReflectionUtils.findMethod(TestB::class.java, "b14", Long::class.java)
            assertTrue(b14.declaringClass == TestB::class.java && b14.name == "b14")
            assertTrue(b14.invoke(null, 14L) as Long == 14L)

            // TODO fix the following case
//            val b15 = ReflectionUtils.findMethod(
//                TestB::class.java,
//                "b15",
//                Int::class.java,
//                String::class.java
//            )
//            assertTrue(b15.declaringClass == TestB::class.java && b15.name == "b15")
//            assertTrue(b15.invoke(null, 15, "b") == "b15")
        } catch (e: Exception) {
            e.printStackTrace()
            fail("failed to reflect: " + e.toString())
        }
    }

    @Test
    fun test_findField() {
        try {
            val objB = TestB()

            // TestA part
            val a1 = ReflectionUtils.findField(TestB::class.java, "a1")
            assertTrue(a1.declaringClass == TestA::class.java && a1.name == "a1")
            a1.set(objB, "a1")
            assertTrue(a1.get(objB) == "a1")

            val a2 = ReflectionUtils.findField(TestB::class.java, "a2")
            assertTrue(a2.declaringClass == TestA::class.java && a2.name == "a2")
            a2.set(objB, 2)
            assertTrue(a2.get(objB) as Int == 2)

            val a3 = ReflectionUtils.findField(TestB::class.java, "a3")
            assertTrue(a3.declaringClass == TestA::class.java && a3.name == "a3")
            a3.set(objB, "a3")
            assertTrue(a3.get(objB) == "a3")

            val a4 = ReflectionUtils.findField(TestB::class.java, "a4")
            assertTrue(a4.declaringClass == TestA::class.java && a4.name == "a4")
            a4.set(objB, 4L)
            assertTrue(a4.get(objB) as Long == 4L)

            val a5 = ReflectionUtils.findField(TestB::class.java, "a5")
            assertTrue(a5.declaringClass == TestA::class.java && a5.name == "a5")
            a5.set(null, "a5")
            assertTrue(a5.get(null) == "a5")

            val a6 = ReflectionUtils.findField(TestB::class.java, "a6")
            assertTrue(a6.declaringClass == TestA::class.java && a6.name == "a6")
            a6.set(null, 6)
            assertTrue(a6.get(null) as Int == 6)

            val a7 = ReflectionUtils.findField(TestB::class.java, "a7")
            assertTrue(a7.declaringClass == TestA::class.java && a7.name == "a7")
            a7.set(null, "a7")
            assertTrue(a7.get(null) == "a7")

            val a8 = ReflectionUtils.findField(TestB::class.java, "a8")
            assertTrue(a8.declaringClass == TestA::class.java && a8.name == "a8")
            a8.set(null, 8L)
            assertTrue(a8.get(null) as Long == 8L)

            // TestB part
            val b1 = ReflectionUtils.findField(TestB::class.java, "b1")
            assertTrue(b1.declaringClass == TestB::class.java && b1.name == "b1")
            a1.set(objB, "b1")
            assertTrue(a1.get(objB) == "b1")

            val b2 = ReflectionUtils.findField(TestB::class.java, "b2")
            assertTrue(b2.declaringClass == TestB::class.java && b2.name == "b2")
            b2.set(objB, 2)
            assertTrue(b2.get(objB) as Int == 2)

            val b3 = ReflectionUtils.findField(TestB::class.java, "b3")
            assertTrue(b3.declaringClass == TestB::class.java && b3.name == "b3")
            b3.set(objB, "b3")
            assertTrue(b3.get(objB) == "b3")

            val b4 = ReflectionUtils.findField(TestB::class.java, "b4")
            assertTrue(b4.declaringClass == TestB::class.java && b4.name == "b4")
            b4.set(objB, 4L)
            assertTrue(b4.get(objB) as Long == 4L)

            val b5 = ReflectionUtils.findField(TestB::class.java, "b5")
            assertTrue(b5.declaringClass == TestB::class.java && b5.name == "b5")
            b5.set(null, "b5")
            assertTrue(b5.get(null) == "b5")

            val b6 = ReflectionUtils.findField(TestB::class.java, "b6")
            assertTrue(b6.declaringClass == TestB::class.java && b6.name == "b6")
            b6.set(null, 6)
            assertTrue(b6.get(null) as Int == 6)

            val b7 = ReflectionUtils.findField(TestB::class.java, "b7")
            assertTrue(b7.declaringClass == TestB::class.java && b7.name == "b7")
            b7.set(null, "b7")
            assertTrue(b7.get(null) == "b7")

            val b8 = ReflectionUtils.findField(TestB::class.java, "b8")
            assertTrue(b8.declaringClass == TestB::class.java && b8.name == "b8")
            b8.set(null, 8L)
            assertTrue(b8.get(null) as Long == 8L)
        } catch (e: Exception) {
            e.printStackTrace()
            fail("failed to reflect: " + e.toString())
        }
    }
}
